sql: look up schemas on database for negative name resolution

thoszhang opened the pull request:
See individual commits for details.

Closes #53301.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53290)
<!-- Reviewable:end -->

lucy-zhang commented:
First 2 commits are ready for review, if you want.

lucy-zhang commented:
Tagging Paul for commit 1. (I also reordered the commits, so the sql-related ones are 2 and 3). This still needs a test for the cleanup on failed restore case, but I think it's ready for a look.

rohany commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -847,7 +848,12 @@ func createImportingDescriptors(
+	log.Eventf(ctx, "starting restore for %d tables", len(mutableTables))
+
+	// Assign new IDs to the database descriptors.
+	if err := rewriteDatabaseDescs(mutableDatabases, details.DescriptorRewrites); err != nil {
```
was this a bug that this wasn't happening before

lucy-zhang replied:
It was happening before (above). I just moved it to a separate function.

rohany commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -923,10 +936,57 @@ func createImportingDescriptors(
+			existingDBsWithNewSchemas := make(map[descpb.ID][]catalog.SchemaDescriptor)
+			for _, sc := range writtenSchemas {
+				parentID := sc.GetParentID()
+				if _, ok := newDatabaseIDs[parentID]; !ok {
```
you could also look at the descRewriteMap if that is present

lucy-zhang replied:
Done.

rohany commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -923,10 +936,57 @@ func createImportingDescriptors(
+				for _, sc := range schemas {
+					db.Schemas[sc.GetName()] = descpb.DatabaseDescriptor_SchemaInfo{ID: sc.GetID()}
+				}
+				// TODO (lucy): It's risky to be incrementing the version outside of the
```
You could do this instead at `publishDescriptors` and use `PublishMultiple` to handle this. I don't think that we need these entries present until we actually turn the tables online

lucy-zhang replied:
Hmm, I'm trying to keep the schema and database descriptor writes in the same transaction. Maybe we should eventually use `PublishMultiple` (or whatever it becomes) at the start of the job too.

rohany commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -950,6 +1010,11 @@ func createImportingDescriptors(
 					}
 					typDesc := desc.(*typedesc.Mutable)
 					typDesc.AddReferencingDescriptorID(table.GetID())
+					// TODO (lucy): I think we should be incrementing the version for the
```
maybe this stuff too

lucy-zhang replied:
Is it valid to add the table descriptors (even if offline) but not add the types they reference at the same time?

rohany replied:
hmm, probably not. We need to have those backreferences there to avoid dropping the type as soon as we add the table descriptor to the database.

rohany commented on pkg/sql/catalog/descs/collection.go:
```diff
@@ -483,22 +483,49 @@ func (tc *Collection) getUserDefinedSchemaVersion(
+		return nil, nil
+	}
+
+	// If we have a schema ID from the database, get the schema descriptor. Since
```
It doesn't seem like there is anything we can do about this without introducing that extra lookup right?

lucy-zhang replied:
Sorry, what do you mean by "this"?

rohany replied:
The false negatives

lucy-zhang replied:
Yeah, I don't think there's a way to completely eliminate this. We could maybe purge old leases for the database descriptor whenever we do it for the schema, but that still doesn't provide a guarantee.

pbardea reviewed (commented):
> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53290)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, @pbardea, and @rohany)

___
*[pkg/ccl/backupccl/restore_job.go, line 946 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFVtC0UoAKWtuXk7OsK-r1-946:-MFXDhjr99pkpbPV6m0R:b-1072zd) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad72e7d95e1d52fb90b3749a2b528e8a068fafc/pkg/ccl/backupccl/restore_job.go#L946)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

you could also look at the descRewriteMap if that is present
</blockquote></details>

I think the rewrite map should be present if this was moved to `publishDescriptors`.
___
*[pkg/ccl/backupccl/restore_job.go, line 967 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFVtC0UoAKWtuXk7OsK-r1-967:-MFXDBynDu2REOY2ZSZI:b-113u7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad72e7d95e1d52fb90b3749a2b528e8a068fafc/pkg/ccl/backupccl/restore_job.go#L967)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

You could do this instead at `publishDescriptors` and use `PublishMultiple` to handle this. I don't think that we need these entries present until we actually turn the tables online
</blockquote></details>

+1 to doing this in `publishDescriptors` - unfortunately we still need to be careful about cleaning this up (like is currently done) in case of an error between marking the descriptors public and actually finishing the job.


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
> +1 to doing this in `publishDescriptors` - unfortunately we still need to be careful about cleaning this up (like is currently done) in case of an error between marking the descriptors public and actually finishing the job.

Hmm, I would really like to not have the database descriptor map get out of sync with the schema namespace entries if at all possible. (It would also make it annoying if in the future, as is my hope, we start enforcing the invariant on writes.) I'm open to this as a short-term solution, but it seems to me like the best way to do this is to also call `PublishMultiple` (or equivalent) at the start of the job to update the type backreferences and the databases (and also creating all the descriptors in the same transaction). Would it be acceptable if we just left the type and database updates the way they are, and come back later to clean them up?

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53290)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @pbardea, and @rohany)

___
*[pkg/ccl/backupccl/restore_job.go, line 967 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFVtC0UoAKWtuXk7OsK-r1-967:-MFXRXPV6s-LU_X1UeRo:b-sklfpv) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad72e7d95e1d52fb90b3749a2b528e8a068fafc/pkg/ccl/backupccl/restore_job.go#L967)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

+1 to doing this in `publishDescriptors` - unfortunately we still need to be careful about cleaning this up (like is currently done) in case of an error between marking the descriptors public and actually finishing the job.
</blockquote></details>

Hmm, right. I missed this when reviewing the schema stuff - but marking the schemas as public at the start of the restore is a bit iffy since they don't have an offline state. With the schema this way, a user might be able to create a table under this schema, and if the restore is cancelled the schema will be deleted from under them.

I think that we'll want to either 1) add OFFLINE capabilities for the type/schema descriptors and then the validation can enforce that the backreferences point to any non-OFFLINE descriptors, or 2) only create the schema and types when marking the tables public.

My current thinking is leaning towards option (1).


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53290)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @pbardea, and @rohany)

___
*[pkg/ccl/backupccl/restore_job.go, line 967 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFVtC0UoAKWtuXk7OsK-r1-967:-MFXUsGm-XF2ZnaRnUxf:bl4rqks) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad72e7d95e1d52fb90b3749a2b528e8a068fafc/pkg/ccl/backupccl/restore_job.go#L967)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Hmm, right. I missed this when reviewing the schema stuff - but marking the schemas as public at the start of the restore is a bit iffy since they don't have an offline state. With the schema this way, a user might be able to create a table under this schema, and if the restore is cancelled the schema will be deleted from under them.

I think that we'll want to either 1) add OFFLINE capabilities for the type/schema descriptors and then the validation can enforce that the backreferences point to any non-OFFLINE descriptors, or 2) only create the schema and types when marking the tables public.

My current thinking is leaning towards option (1).
</blockquote></details>

Yeah, I am also in favor of adding an offline state for all the descriptors.

If we don't end up doing that in 20.2 for whatever reason, I think we should at least stop deleting schemas if the restore was canceled and it had tables added in the meantime, like we do with databases.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Can you help me better understand how you're thinking they'll get out of sync now?

> Reviewed 3 of 3 files at r2, 4 of 4 files at r5, 1 of 1 files at r6, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-:-MFXZ2K55bPCV_iNPhDb:bqsd232)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang and @rohany)

___
*[pkg/ccl/backupccl/restore_job.go, line 1013 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFVtC0UoAKWtuXk7OsK-r1-1013:-MFXWyV0CdjwXpBntJ0n:b-bd18zn) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad72e7d95e1d52fb90b3749a2b528e8a068fafc/pkg/ccl/backupccl/restore_job.go#L1013)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Is it valid to add the table descriptors (even if offline) but not add the types they reference at the same time?
</blockquote></details>

Seems like we're not doing the full validation. Maybe we should be. I've been thinking that in `CommitSQLTransaction` we should go validate all of the modified descriptors in the collection. Sort of by definition every reference should already be in the `Collection` so it should just be an in-memory operation and I suspect it would catch bugs. 

We should be doing the same sort of thing in restore. 
___
*[pkg/ccl/backupccl/restore_job.go, line 945 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFXVxGQBpVqYSdaVFqn:-MFXVxGQBpVqYSdaVFqo:bx9azmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/965f2d798c065454695d920ee115b1435d428c1b/pkg/ccl/backupccl/restore_job.go#L945)):*
> ```Go
> 			// Write the updated databases.
> 			for dbID, schemas := range existingDBsWithNewSchemas {
> 				log.Infof(ctx, "writing schema entries to database %d", dbID)
> ```

nit: you've got then number of entries, might as well add it
___
*[pkg/ccl/backupccl/restore_job.go, line 952 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFXW8tZ8jRt45a_f6Xc:-MFXW8tZ8jRt45a_f6Xd:b11gces) ([raw file](https://github.com/cockroachdb/cockroach/blob/965f2d798c065454695d920ee115b1435d428c1b/pkg/ccl/backupccl/restore_job.go#L952)):*
> ```
> 					// TODO (lucy): Push map initialization to when we create these
> 					// descriptors.
> ```
I think I even prefer not directly accessing the map but rather having an `AddSchema`/`UpsertSchema` method or something that will lazily initialize it. We're going to pull all of these mutations under the interface hopefully eventually. 
___
*[pkg/ccl/backupccl/restore_job.go, line 962 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFXWQkM068MBdWHBxxI:-MFXWQkNBKqKfPEglE7z:b-z7p5w7) ([raw file](https://github.com/cockroachdb/cockroach/blob/965f2d798c065454695d920ee115b1435d428c1b/pkg/ccl/backupccl/restore_job.go#L962)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 				// "mutable" descriptors right out of the store. We should be fine since
> 				// we're updating each of these databases exactly once, but there's
> 				// nothing preventing us from mistakenly incrementing the version the
> 				// wrong number of times in the transaction.
> ```
</details>

In theory, if we did it right, the whole `MaybeIncrementVersionThing` which looks at the `ClusterVersion` prevents it.

I don't feel that the problem is the call to `MaybeIncrementVersion` but rather the fact that we're not using a `descs.Collection` to manage resolving mutable descriptors. In my dream world we won't hit the database when we re-resolve a mutable descriptor. 
___
*[pkg/ccl/backupccl/restore_planning.go, line 715 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFXXVTs75ewCE0DXsbo:-MFXXVTs75ewCE0DXsbp:b-42otop) ([raw file](https://github.com/cockroachdb/cockroach/blob/965f2d798c065454695d920ee115b1435d428c1b/pkg/ccl/backupccl/restore_planning.go#L715)):*
> ```Go
> }
> 
> // rewriteDatabaseDescs rewrites all ID's in the input slice of
> ```

This feels like the class of pure function that could be nice to unit test. (not asking for that here, just a passing thought on the accumulated lack of unit testing which one day we may want to shift our attention towards). 
___
*[pkg/sql/catalog/descs/collection.go, line 511 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFXYM7p1LfF6PMf9Wh_:-MFXYM7p1LfF6PMf9Wha:b6i3uwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/deb619a44b42809187b9d3aa6392e5ae0f746d8f/pkg/sql/catalog/descs/collection.go#L511)):*
> ```Go
> 	}
> 
> 	// If we have a schema ID from the database, get the schema descriptor. Since
> ```

This seems fine, right? It can only happen if we just added or modified the schema and haven't yet waited for leases to drain on the old version of the database. We're not violating any contracts because we know the `CREATE SCHEMA` or `ALTER SCHEMA` hasn't yet returned to the client. 

If we're really worried about it in the local node, we could store a version of the database in the schema descriptor and ensure that the leased database in the `lease.Manager` is of a high enough version, but I'm still not clear on the hazard.


<!-- Sent from Reviewable.io -->

pbardea reviewed (approved):
Approach of the first commit generally looks good to me with the caveat that we'll want to clean it up according to https://github.com/cockroachdb/cockroach/issues/53375.

A test for the cleanup would be nice - backup tests have simulated a bad backup by removing the SSTs (https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/backupccl/backup_test.go#L1691).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-:-MFXhHLZ4F5EAe1x3qsd:b-tn1n2t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang and @rohany)

___
*[pkg/ccl/backupccl/restore_job.go, line 967 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFVtC0UoAKWtuXk7OsK-r1-967:-MFXcnwz0jj7bT9fQuRq:b-txvqhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad72e7d95e1d52fb90b3749a2b528e8a068fafc/pkg/ccl/backupccl/restore_job.go#L967)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Yeah, I am also in favor of adding an offline state for all the descriptors.

If we don't end up doing that in 20.2 for whatever reason, I think we should at least stop deleting schemas if the restore was canceled and it had tables added in the meantime, like we do with databases.
</blockquote></details>

Thanks for filing #53375. I'm fine with leaving this as is for now and cleaning it up with the OFFLINE fix in a follow up.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> Can you help me better understand how you're thinking they'll get out of sync now?

Sorry, I'm not sure what you mean. Are you responding to this?

> Hmm, I would really like to not have the database descriptor map get out of sync with the schema namespace entries if at all possible. 

(This is kind of a confusingly threaded comment, also - the rest of the discussion I had with Paul inline). We were just discussing the hypothetical possibility of moving the schema map updates toward the end of the job, instead of in the same transaction as when the database descriptors get written early on, and I said I didn't really want to do it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-:-MFXl3Iw4ZrFco_ZtyWm:b5zkn0o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @rohany)

___
*[pkg/ccl/backupccl/restore_job.go, line 967 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFVtC0UoAKWtuXk7OsK-r1-967:-MFXcnI9ENoeSsIffSJQ:b-5vhc49) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad72e7d95e1d52fb90b3749a2b528e8a068fafc/pkg/ccl/backupccl/restore_job.go#L967)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Yeah, I am also in favor of adding an offline state for all the descriptors.

If we don't end up doing that in 20.2 for whatever reason, I think we should at least stop deleting schemas if the restore was canceled and it had tables added in the meantime, like we do with databases.
</blockquote></details>

I opened #53375 to track using offline states for the new descriptors. I'll also open separate issues for (1) the orphaned child objects under a deleted schema after we clean up a failed restore, and (2) whatever we want to do to make updating descriptors safer than just reading them straight from the store and writing them (which Andrew and I are discussing on another thread).
___
*[pkg/ccl/backupccl/restore_job.go, line 1013 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFVtC0UoAKWtuXk7OsK-r1-1013:-MFXgvp4E_5U80CXDRZR:bb7blgk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad72e7d95e1d52fb90b3749a2b528e8a068fafc/pkg/ccl/backupccl/restore_job.go#L1013)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Seems like we're not doing the full validation. Maybe we should be. I've been thinking that in `CommitSQLTransaction` we should go validate all of the modified descriptors in the collection. Sort of by definition every reference should already be in the `Collection` so it should just be an in-memory operation and I suspect it would catch bugs. 

We should be doing the same sort of thing in restore. 
</blockquote></details>

Yeah, I think that makes sense. I've been writing up a meta-issue about all these RESTORE issues and will add this.
___
*[pkg/ccl/backupccl/restore_job.go, line 952 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFXW8tZ8jRt45a_f6Xc:-MFXb_elEra70Qf3polk:b-hgylkl) ([raw file](https://github.com/cockroachdb/cockroach/blob/965f2d798c065454695d920ee115b1435d428c1b/pkg/ccl/backupccl/restore_job.go#L952)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 					// TODO (lucy): Push map initialization to when we create these
> 					// descriptors.
> ```
I think I even prefer not directly accessing the map but rather having an `AddSchema`/`UpsertSchema` method or something that will lazily initialize it. We're going to pull all of these mutations under the interface hopefully eventually. 
</blockquote></details>

Yeah, that's fair. I just took out the comment, since we have many glaring examples where we should hide the `Schemas` manipulation under an interface.
___
*[pkg/ccl/backupccl/restore_job.go, line 962 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFXWQkM068MBdWHBxxI:-MFXcwWX52NBLCBusnjI:b6yogl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/965f2d798c065454695d920ee115b1435d428c1b/pkg/ccl/backupccl/restore_job.go#L962)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 				// "mutable" descriptors right out of the store. We should be fine since
> 				// we're updating each of these databases exactly once, but there's
> 				// nothing preventing us from mistakenly incrementing the version the
> 				// wrong number of times in the transaction.
> ```
In theory, if we did it right, the whole `MaybeIncrementVersionThing` which looks at the `ClusterVersion` prevents it.

I don't feel that the problem is the call to `MaybeIncrementVersion` but rather the fact that we're not using a `descs.Collection` to manage resolving mutable descriptors. In my dream world we won't hit the database when we re-resolve a mutable descriptor. 
</blockquote></details>

Yeah, I agree with you entirely on this. My original comment wasn't very clear, but my point was that the `ClusterVersion` itself is going to become incorrect if we read the descriptor directly from the store, call `MaybeIncrementVersion`, and write it back to the store multiple times in this transaction. I updated the comments to hopefully make this more clear.
___
*[pkg/sql/catalog/descs/collection.go, line 511 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-MFXYM7p1LfF6PMf9Wh_:-MFXjdv85jFS6DKsP4GQ:b-ge5v22) ([raw file](https://github.com/cockroachdb/cockroach/blob/deb619a44b42809187b9d3aa6392e5ae0f746d8f/pkg/sql/catalog/descs/collection.go#L511)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This seems fine, right? It can only happen if we just added or modified the schema and haven't yet waited for leases to drain on the old version of the database. We're not violating any contracts because we know the `CREATE SCHEMA` or `ALTER SCHEMA` hasn't yet returned to the client. 

If we're really worried about it in the local node, we could store a version of the database in the schema descriptor and ensure that the leased database in the `lease.Manager` is of a high enough version, but I'm still not clear on the hazard.
</blockquote></details>

Yeah, it's fine. This comment is probably too alarming, I added a bit toward the end clarifying that it's fine.


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
For what it's worth, I've gathered up all the improvements discussed in this review (and others) into a meta-issue for future work, some of which I expect to be done for 20.2. #53378

lucy-zhang commented:
> A test for the cleanup would be nice - backup tests have simulated a bad backup by removing the SSTs (https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/backupccl/backup_test.go#L1691).

I tacked on a check toward the end of the existing test. Let me try to expand on it a bit.

Commit:
sql: look up schemas on database for negative name resolution

Previously, when resolving a 2-part object name like `x.y`, the
ambiguity in whether `x` is a schema in the current database or a
database whose public schema contains `y` meant that if `x` were a
database, we would need to try to resolve it as a schema by reading from the
store every time the name was used. This PR introduces the use of the
database's map of schema names to IDs, which is complete and thus allows
returning early with a negative result if a schema is not in the
database. With this change, in the common case where all descriptors
have been leased, no round-trips are required for name resolution for
2-part names.

Release note (performance improvement): Fixes a performance regression
introduced in 20.2 alpha releases where the use of 2-part names to
specify tables or types would incur extra reads from the store on every
transaction.

Commit:
backupccl: rewrite database schema maps on restore

This commit updates the restore job to update the parent
`DatabaseDescriptor.Schemas` map whenever a schema is created (in the
resumer) or dropped (during cleanup). This enables the correct
functioning of schema name resolution for restored descriptors.

Release note: None

Commit:
sql: write updated parent db when reparenting schema

Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though I hope we can resolve most of these major concerns before 20.2 goes out the door. Here's my understanding of the list:


* Utilize a desc.Collection for handling descriptors so we don't read a mutated descriptor from the store 
    * Increment type versions when adding backreferences
* Add the schemas and types in an offline state (#53375)
    * Test these scenarios involving rolling back restores which utilize these types

Bonus:
* Add a validation stage prior to committing to validate the integrity of everything being written


Does that line up with what you're thinking?

> Reviewed 5 of 5 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53290#-:-MFYCVSv7i7MPx2eM5lH:b-ct7mlv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @rohany)



<!-- Sent from Reviewable.io -->

rohany commented:
I don't really have much to add, but the discussions above all sound right, and I'm also a proponent of adding an OFFLINE state to these descriptors so that we have more flexibility while mucking around with them during restores.

lucy-zhang commented:
TFTRs!

bors r+

lucy-zhang commented:
@ajwerner Yeah, I think I'm on board with that prioritized list. I could maybe live without fixing all the descriptor accesses in full generality for 20.2, but in principle it shouldn't be too hard.

Here's that meta-issue for RESTORE improvements again (posting where it's easier to see): #53378
